### PR TITLE
AES-GCM x86_64 MSVC ASM: XMM6-15 are non-volatile

### LINKED ISSUE #6608: [Bug]:  aes_gcm_asm.asm doesn't support Windows x86-64 calling convention
### Version

5.6.3

### Description
 aes_gcm_asm.asm doesn't support Windows x86-64 calling convention

[according](https://learn.microsoft.com/en-us/cpp/build/x64-software-conventions?view=msvc-170#register-volatility-and-preservation) to MS convention, registers XMM6-15 are nonvolatile. 

see my [previous](https://github.com/wolfSSL/wolfssl/issues/5190) report for details please

for example, clang can use XMM6-15 registers for initialization constants

### PR DESCRIPTION
**Pull request by SparkiDev**
# Description

Put XMM6-15, when used, on the stack at start of function and restore at end of function.

Fixes #6608

# Testing

Standard

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: wolfcrypt/src/aes_gcm_asm.asm [hunk 1]
```diff
@@ -108,14 +108,24 @@ AES_GCM_encrypt PROC
...
         mov	rax, r9
-        mov	r8, QWORD PTR [rsp+96]
-        mov	r9d, DWORD PTR [rsp+104]
-        mov	r11d, DWORD PTR [rsp+112]
-        mov	ebx, DWORD PTR [rsp+120]
-        mov	r14d, DWORD PTR [rsp+128]
-        mov	r15, QWORD PTR [rsp+136]
-        mov	r10d, DWORD PTR [rsp+144]
-        sub	rsp, 160
+        mov	r8, QWORD PTR [rsp+256]
```
Comment (ilka1999) [inline wolfcrypt/src/aes_gcm_asm.asm:111]: why +256 ?
it should be +96

File: wolfcrypt/src/aes_gcm_asm.asm [hunk 2]
```diff
@@ -1960,15 +1980,25 @@ AES_GCM_decrypt PROC
...
-        mov	r8, QWORD PTR [rsp+104]
-        mov	r9d, DWORD PTR [rsp+112]
-        mov	r11d, DWORD PTR [rsp+120]
-        mov	ebx, DWORD PTR [rsp+128]
-        mov	r14d, DWORD PTR [rsp+136]
-        mov	r15, QWORD PTR [rsp+144]
-        mov	r10d, DWORD PTR [rsp+152]
-        mov	rbp, QWORD PTR [rsp+160]
-        sub	rsp, 168
+        mov	r8, QWORD PTR [rsp+264]
```
Comment (ilka1999) [inline wolfcrypt/src/aes_gcm_asm.asm:1983]: again why +264 ?

File: wolfcrypt/src/aes_gcm_asm.asm [hunk 3]
```diff
@@ -108,14 +108,24 @@ AES_GCM_encrypt PROC
...
-        mov	r10d, DWORD PTR [rsp+144]
-        sub	rsp, 160
+        mov	r8, QWORD PTR [rsp+256]
+        mov	r9d, DWORD PTR [rsp+264]
+        mov	r11d, DWORD PTR [rsp+272]
+        mov	ebx, DWORD PTR [rsp+280]
+        mov	r14d, DWORD PTR [rsp+288]
+        mov	r15, QWORD PTR [rsp+296]
+        mov	r10d, DWORD PTR [rsp+304]
+        vmovdqu	OWORD PTR [rsp+160], xmm6
```
Comment (ilka1999) [inline wolfcrypt/src/aes_gcm_asm.asm:118]: why AVX version instead of SSE (movdqu) ?

File: wolfcrypt/src/aes_gcm_asm.asm [hunk 4]
```diff
@@ -5102,10 +5176,12 @@ AES_GCM_encrypt_final_aesni PROC
...
         mov	r9, rdx
-        mov	r11d, DWORD PTR [rsp+64]
-        mov	r12, QWORD PTR [rsp+72]
-        mov	r14, QWORD PTR [rsp+80]
-        sub	rsp, 16
+        mov	r11d, DWORD PTR [rsp+96]
+        mov	r12, QWORD PTR [rsp+104]
+        mov	r14, QWORD PTR [rsp+112]
+        vmovdqu	OWORD PTR [rsp+16], xmm6
+        vmovdqu	OWORD PTR [rsp+32], xmm7
```
Comment (ilka1999) [inline wolfcrypt/src/aes_gcm_asm.asm:5183]: what about xmm8-14 ?

**Review by ilka1999 — COMMENTED** [review]
Common mistake is the wrong offset for parameters which passed by stack.
These parameters are loaded before stack reservation for local purposes, so their offset should be the same, as it was before stack usage increase.

File: wolfcrypt/src/aes_gcm_asm.asm (hunk 1 above)
Comment (SparkiDev) [reply wolfcrypt/src/aes_gcm_asm.asm:111]: I use 20 64-bit words of stack in the function for temporary storage.

File: wolfcrypt/src/aes_gcm_asm.asm (hunk 3 above)
Comment (SparkiDev) [reply wolfcrypt/src/aes_gcm_asm.asm:118]: Good point.
Generated code and it didn't know to produce SSE2 only code.
Fixed this.

File: wolfcrypt/src/aes_gcm_asm.asm (hunk 2 above)
Comment (SparkiDev) [reply wolfcrypt/src/aes_gcm_asm.asm:1983]: As above, stack used for temporary storage.

File: wolfcrypt/src/aes_gcm_asm.asm (hunk 4 above)
Comment (SparkiDev) [reply wolfcrypt/src/aes_gcm_asm.asm:5183]: Fixed generating code.

File: wolfcrypt/src/aes_gcm_asm.asm (hunk 1 above)
Comment (ilka1999) [reply wolfcrypt/src/aes_gcm_asm.asm:111]: Sean, 

It is loading parameters from the stack

Let's calculate offset for 5-th parameter:

1. first four parameters  = 32 bytes
2. return address = 8 bytes
3. seven non-volatile registers saving = 56 bytes

totally 96 bytes
so, 5-th parameter has offset 96 from RSP

**Review by ilka1999 — CHANGES_REQUESTED** [review]
at least you need

1. change commands for all XMM saves from vmovdqu to movdqu
2. check the parameter offsets, they should not be changed unless there are no new stack modifications between the start of the function and the loading of the parameters.

be careful about addressing parameters from the middle of the code, it should be increased if such addressing occurs after new stack modifications

File: wolfcrypt/src/aes_gcm_asm.asm (hunk 1 above)
Comment (SparkiDev) [reply wolfcrypt/src/aes_gcm_asm.asm]: Looks like I confused myself and treated the xmm registers to save as parameters.
Fix up now.

**Comment by SparkiDev** [comment]
Fixed generation to put on stack and take off stack in the right place.

**Comment by ilka1999** [comment]
Ok

Last version is working with my friend's tests

But, ASM code is not perfect:
1. there are some oddish lines
2. no proc frames

**Comment by SparkiDev** [comment]
Hi @ilka1999

1. Let me know what is odds in the lines. Specifics or generally.
2. Are you suggesting adding 'FRAMES' to the PROC lines?

Thanks,
Sean

**Comment by ilka1999** [comment]
IMHO:

1. the code still uses vmovdqu instead of movdqu
2. savings of XMM registers which does not changed
3. allocates stack for temporary local variables in the middle of body

I did not check the stack alignment of the code, with alignment you can use ALIGNED loading/store instead of UNALIGNED

**Comment by SparkiDev** [comment]
Hi @ilka1999,

Thanks for the feedback!

1. I've pushed an update to fix this.
2.This has also been fixed.
3. The temporary stack allocations are not something I intend to change right now.

Note that the stack is not always aligned as I would like. On newer processors, unaligned moves are the same speed as the aligned moves. I won't be making changes for this but thank you for bringing it up.

Sean

**Comment by SparkiDev** [comment]
retest this please

**Comment by SparkiDev** [comment]
retest this please

**Comment by ilka1999** [comment]
Tested successfully

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**